### DISCUSSION TITLE
Bad refresh token when refreshing an user-to-server token

### DISCUSSION THREAD

**Discussion by fuegoio** [community/community#24745] [Apps, API and Webhooks]
<p>Hi everyone,</p>
<p>We created a Github App and currently using it for server-to-server requests and user-to-server requests. We opted-in the Expiring token feature, but after 8 hours or so, when the access token is expired, it is impossible to refresh the token according to the documentation.</p>
<p>This is what we get from our <code>POST</code> to refresh our access token:</p>
<pre><code class="lang-auto">Error: bad_refresh_token
Description: The refresh token passed is incorrect or expired.
</code></pre>
<p>The documentation says that the refresh token should last 6 month, but we do think that this isn’t the case.</p>
<p>What can we do to fix this ?</p>
<p>Thanks a lot,</p>
<p>Alexis</p>

**Comment by laszlocph** [comment]
<p>I have the same :man_shrugging:</p>
<p>turned it off in the app settings as it is a new beta feature</p>

**Reply by Black-toot** [reply]
https://www.lemon8-app.com/discover/837246?region=sg&pid=website_seo_from_sug_fof#8080000000055__

**Comment by francisfuzz** [comment, marked as answer]
<p><a class="mention" href="/fuegoio">@fuegoio</a> &amp; <a class="mention" href="/laszlocph">@laszlocph</a> –– thanks for reporting this. We’ve escalated this to our engineering team and will follow up here when we have an update from them; we don’t have a timeline for when this will be addressed. :bowing_man:</p>

**Reply by Black-toot** [reply]
37.35.79.24/?data:text/html,#652986090#

**Comment by hatboysam** [comment]
<p>I am seeing the same thing … I can’t exchange refresh tokens for access tokens.</p>

**Comment by hatboysam** [comment]
<p>Does this mean this is a confirmed bug?  Is it recommended that GitHub App developers move to non-expiring tokens?</p>

**Comment by francisfuzz** [comment]
<p>:wave: <a class="mention" href="/hatboysam">@hatboysam</a>: I reported this to our engineering team last week; they’re looking into it and the behavior is not a confirmed bug at this time. I’m awaiting their recommendation as a part of their investigation and will follow up here once I hear more.</p>

**Comment by francisfuzz** [comment]
<p>:wave: Hello everyone! Francis here from the GitHub Support team. Our engineering deployed a new update so this issue should be resolved.</p>
<p><a class="mention" href="/fuegoio">@fuegoio</a>, <a class="mention" href="/laszlocph">@laszlocph</a>, <a class="mention" href="/hatboysam">@hatboysam</a>: I’m wondering if you all are still able to reproduce this error when refreshing an user-to-server token? :thought_balloon:</p>
<p>If so, sharing the full request-response pair of the executed request would help us investigate (please be sure to redact any sensitive information like authorization headers, credentials, and similar data). :bowing_man:</p>

**Comment by hatboysam** [comment]
<p><a class="mention" href="/francisfuzz">@francisfuzz</a> this is still happening, here’s what I got from my logs.  It’s just the response, I’ll try to get the full request although hopefully you can look it up:</p>
<pre><code class="lang-auto">[1] &gt;  Status code: 200
[1] &gt;  Response headers {"date":"Sat, 29 Aug 2020 09:33:23 GMT","content-type":"application/x-www-form-urlencoded; charset=utf-8","transfer-encoding":"chunked","connection":"close","server":"GitHub.com","status":"200 OK","vary":"X-PJAX, Accept-Encoding, Accept, X-Requested-With, Accept-Encoding","etag":"W/\"a5eb409e6abcc06043c22fc0c114a1cb\"","cache-control":"max-age=0, private, must-revalidate","strict-transport-security":"max-age=31536000; includeSubdomains; preload","x-frame-options":"deny","x-content-type-options":"nosniff","x-xss-protection":"1; mode=block","referrer-policy":"origin-when-cross-origin, strict-origin-when-cross-origin","expect-ct":"max-age=2592000, report-uri=\"https://api.github.com/_private/browser/errors\"","content-security-policy":"default-src 'none'; base-uri 'self'; block-all-mixed-content; connect-src 'self' uploads.github.com www.githubstatus.com collector.githubapp.com api.github.com www.google-analytics.com github-cloud.s3.amazonaws.com github-production-repository-file-5c1aeb.s3.amazonaws.com github-production-upload-manifest-file-7fdce7.s3.amazonaws.com github-production-user-asset-6210df.s3.amazonaws.com cdn.optimizely.com logx.optimizely.com/v1/events wss://alive.github.com; font-src github.githubassets.com; form-action 'self' github.com gist.github.com; frame-ancestors 'none'; frame-src render.githubusercontent.com; img-src 'self' data: github.githubassets.com identicons.github.com collector.githubapp.com github-cloud.s3.amazonaws.com *.githubusercontent.com; manifest-src 'self'; media-src 'none'; script-src github.githubassets.com; style-src 'unsafe-inline' github.githubassets.com; worker-src github.com/socket-worker.js gist.github.com/socket-worker.js","x-github-request-id":"E503:DB62:514CD51:78C2253:5F4A20E3"}    
[1] &gt;  Response data "error=bad_refresh_token&amp;error_description=The+refresh+token+passed+is+incorrect+or+expired.&amp;error_uri=https%3A%2F%2Fdocs.github.com%2Fapps%2Fmanaging-oauth-apps%2Ftroubleshooting-oauth-app-access-token-request-errors%2F%23bad-verification-code"
</code></pre>
<p>I think the piece you’re most interested in is:</p>
<pre><code class="lang-auto">"x-github-request-id":"E503:DB62:514CD51:78C2253:5F4A20E3"
</code></pre>
<p>Edit: the request looks like this:</p>
<pre><code class="lang-auto">[1] &gt;  Request: HTTP post: "https://github.com/login/oauth/access_token?client_id=Iv1.3bfe017ea9365f15&amp;client_secret=REDACTED&amp;grant_type=refresh_token&amp;refresh_token=REDACTED"
[1] &gt;  Request Headers: undefined
[1] &gt;  Request Body: undefined
</code></pre>

**Comment by laszlocph** [comment]
<p>I opted out of the “feature” so can’t say.</p>

**Comment by hatboysam** [comment]
<p>I just tried again and it <em>seems</em> to be working today … is it possible the fix wasn’t rolled out when I last tried?</p>

**Comment by francisfuzz** [comment]
<p><a class="mention" href="/hatboysam">@hatboysam</a> - That’s interesting behavior to report! Now that it’s working, if anything changes in the next week, please share an updated <code>X-GitHub-Request-Id</code> key-value pair and I can forward that over to our engineers for further review. :+1:</p>

**Comment by nlecoy** [comment]
<p>Hi <a class="mention" href="/francisfuzz">@francisfuzz</a>,<br>
Just had the same issue when trying to refresh a user-to-server token. Here is my <code>X-GitHub-Request-Id</code>: <code>D9E0:2FCD:478762B:49A3AE1:60617F28</code>.</p>
<p>I had to opt out the feature to solve this issue.</p>

**Comment by francisfuzz** [comment]
<p><a class="mention" href="/nlecoy">@nlecoy</a> - Thanks for sharing that with us – I’ve forwarded it over to our engineering team for investigation. There isn’t a timeline for when this will be resolved, and I’ll loop back here as soon as we hear more from them.</p>

**Comment by mikeparker** [comment]
<p>I’m also seeing this. After 8 hours the refresh token no longer works, with a bad_refresh_token error. I’ve checked using the refresh token to get a new access token within 8 hours and it works fine, creating new access tokens (which then work), again the refresh token expiry time is in 6 months so its surprising that it’s marked as invalid or expired. (It can’t be invalid as it works fine before 8 hours is up).</p>
<p>Please let me know what I can do to help debug / diagnose, it’s pretty simple to reproduce.</p>

**Comment by francisfuzz** [comment]
<p>:wave: <a class="mention" href="/nlecoy">@nlecoy</a> &amp; <a class="mention" href="/mikeparker">@mikeparker</a> – thanks again for sharing these details with us. One of our engineers triaged this and let us know they haven’t been able to reproduce this. However, they’re keen on receiving any additional information for them to investigate further. Specifically, which application(s) are you respectively using, and are you able to reliably reproduce this outcome? I can relay that over to the team, though I’d like to caution against sharing any private or sensitive information before sharing it on this thread – having step-by-steps that reliably reproduces the behavior would help them, especially <code>X-GitHub-Request-Id</code> values for each respective request that returns this error.</p>

**Comment by xiaoxinghu** [comment]
<p>I can still see the problem exactly as <a class="mention" href="/mikeparker">@mikeparker</a> described above. <strong>Refresh token works as expected within the expiration date, but not after that point.</strong> The problem with reproducing it is you have to wait for 8 hours. I will try to catch some more info tomorrow since I just refreshed my token successfully.</p>

**Comment by backrightupcom** [comment]
<p>We are seeing the same issue - if we receive a access token and refresh token and then 30 minutes later try to refresh the access token we get the same issue.</p>
<p>X-GitHub-Request-Id=C814:6EA7:1372561:155E5B2:609DCCEB</p>
<p>Here is our code:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/9/1/915619690e666b348080d99cf74fdab1d7071f71.png" data-download-href="/uploads/short-url/kJHHoU7DDJv73yqB2xXhdRO3Oet.png?dl=1" title="image" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/81234491/181086541-fe33ebdb-9bb5-41eb-bc9b-ed04657235db.png" alt="image" data-base62-sha1="kJHHoU7DDJv73yqB2xXhdRO3Oet" width="690" height="198" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/9/1/915619690e666b348080d99cf74fdab1d7071f71_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">817×235 22.7 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by backrightupcom** [comment]
<p>I was encoding the refreshtoken. Thats a no no.</p>

**Comment by dmuth** [comment]
<p>Hello!  I found this thread during a Google Search and I think it may be happening to me.</p>
<p>I am doing a self-study of OAuth2 and have been using the Postman desktop client to OAuth myself to a few different sites so I can play around with their APIs.  I’ve been able to authenticate with sites like Imgur and DeviantArt so far, but have been running into difficulties with GitHub.</p>
<p>Like others in this thread, the issue is using my refresh_token to get an access_token.  I’ve been trying for the last couple of hours, and have consistently gotten the bad_refresh_token respond from GitHub’s endpoint.</p>
<p>I’ve highlighted-relevant parts of the request/response (and blurred out secrets):</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/a/9/a98c739f3551e30a9c13751ffe03946a060f29d6.png" data-download-href="/uploads/short-url/obTBK7UmNicAfdM0eB463hrYrsO.png?dl=1" title="CleanShot 2021-06-08 at 19.34.52@2x" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/374060/181086545-a2c8d72d-09e6-4dc0-87ed-25773cd6afb5.png" alt="CleanShot 2021-06-08 at 19.34.52@2x" data-base62-sha1="obTBK7UmNicAfdM0eB463hrYrsO" width="682" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/a/9/a98c739f3551e30a9c13751ffe03946a060f29d6_2_682x500.png, https://aws1.discourse-cdn.com/github/optimized/3X/a/9/a98c739f3551e30a9c13751ffe03946a060f29d6_2_1023x750.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/3X/a/9/a98c739f3551e30a9c13751ffe03946a060f29d6_2_1364x1000.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/a/9/a98c739f3551e30a9c13751ffe03946a060f29d6_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">CleanShot 2021-06-08 at 19.34.52@2x</span><span class="informations">2434×1782 561 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Also, if it helps GitHub engineers in their searching, the X-GitHub-Request-Id n the response is C092:1029:105A80:1F3352:60BFFE7D.</p>
<p>Thanks!</p>
<p>– Doug</p>

**Comment by coryvirok** [comment]
I'm seeing this happen with an access token that has *not* expired but only after I update the github app installation.

E.g.

1. Authenticate and store the token + refresh token as part of the `code=` oauth flow
2. Install the github app and redirect back to my app install setup callback (note: I have oauth during app install disabled)
3. User is redirected back to my app and all is well
4. User updates the github app installation to add/remove repos from the app and is redirected back to my install setup callback
5. In the app install setup callback I use the user token from step 1 to check which installations the user has access to. Octokit determines that the token is not valid so it uses the refresh token from step 1 to get a new one
6. Octokit fails to get the user's installations with the error `bad_refresh_token`

`x-github-request-id`: `FF49:7A0D:D29389:DA15F0:63BCA7B5`

**Comment by tim-thacker-nullify** [comment]
I'm still getting this error.
Not using any SDK, just the Golang HTTP client.
Following exactly what it says to do in the docs here https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/refreshing-user-access-tokens.

Still getting the bad_refresh_token error, "The refresh token passed is incorrect or expired."

```
CLIENT_SECRET=xxxxxxxxxx...
CLIENT_ID=Iv1.xxxxxxxxxx...
REFRESH_TOKEN=ghr_xxxxxxxxxx...
URL="https://github.com/login/oauth/access_token?grant_type=refresh_token&client_id=${CLIENT_ID}&client_secret=${CLIENT_SECRET}&refresh_token=${REFRESH_TOKEN}"
curl -X POST -H "Accept: application/json" $URL
```

**Comment by koslib** [comment]
I am here just to say this is indeed an issue still. @francisfuzz is there a timeline for this by the github team?

IMO this issue is marked as "answered" however it's definitely not.

**Comment by db-lyon** [comment]
I'm facing this issue too. I can't believe it's been 3 years. 👎

**Comment by depombo** [comment]
ran into the same issue

**Comment by rafidRD** [comment]
I am also facing the same problem... glitch in the matrix

**Comment by doei** [comment]
@francisfuzz is this still on your radar? The steps to reproduce are really easy with a GitHub App:

1. Have a Github App that enables User-to-server token expiration
2. Create a user access token using the [web application flow](https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/generating-a-user-access-token-for-a-github-app#using-the-web-application-flow-to-generate-a-user-access-token)
2. Wait 8 hours for the user access token to expire
3. Use the refresh token to get a new access token as per described here: [Refreshing a user access token with a refresh token](https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/refreshing-user-access-tokens#refreshing-a-user-access-token-with-a-refresh-token)
4. This will never work, whether you try via octokit, or Postman...

**Reply by doei** [reply]
One x-github-request-id of my own request experiencing the problem: DA1F:A6040:508199:5144CC:66268CF4

**Comment by iki** [comment]
We have the same issue when trying to refresh the access token immediately after obtaining it (sign-in), i.e. we didn't have to wait for access token to expire (in 8h).

`auth.token.refresh (2024-05-31T10:20:30.729Z): [200] Unexpected token 'e', "error=bad_"... is not valid JSON: error=bad_refresh_token&error_description=The+refresh+token+passed+is+incorrect+or+expired.&error_uri=https%3A%2F%2Fdocs.github.com%2Fapps%2Fmanaging-oauth-apps%2Ftroubleshooting-oauth-app-access-token-request-errors%2F%23bad-verification-code @ https://github.com/login/oauth/access_token?client_id=xxx&client_secret=xxx&grant_type=refresh_token&refresh_token=ghr_...`

```ts
const tokenUrl = 'https://github.com/login/oauth/access_token'

const tokenEndpointOptions = {
  headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
  method: 'POST',
}
const url = getSearchUrl(tokenUrl, {
      client_id: clientId,
      client_secret: clientSecret,
      grant_type: 'refresh_token',
      refresh_token: token.refreshToken!,
    })

const data = await fetchJson(url, tokenEndpointOptions)
```

**Reply by Megakuul** [reply]
I experience the same behavior when using `golang.org/x/oauth2` package. When I create a new oauth2 `TokenSource` immediately after acquiring the refresh token, the token request fails:
```go
tokenSource := routeCtx.OAuthConfig.TokenSource(transportCtx, &oauth2.Token{
	RefreshToken: userDoc.RefreshToken,
})

token, err := tokenSource.Token() // fails immediately (bad_refresh_token&error_description=The+refresh+token+passed+is+incorrect+or+expired)
```

**Comment by shapled** [comment]
Repeatedly using the refresh_token can lead to the bad_refresh_token issue.

**Comment by Sisil69** [comment]
Telegram bot for showing a tell

**Comment by vedant-contentstack** [comment]
Any updates on this issue??

**Comment by allisonee** [comment]
+1 Came here to say that we're still getting this same issue. Any updates from the github team?

Specifically, we get a valid refresh_token that expires in 6 months, but after the provider_token expires (x hours) refreshing the token from the [docs](https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/refreshing-user-access-tokens#refreshing-a-user-access-token-with-a-refresh-token) returns the error `The refresh token passed is incorrect or expired`

Refreshing within the x hours works though, so hard to repro as you've got to catch the issue after that.

**Comment by stefdelec** [comment]
Still  happening and I don't understand the issue.

**Comment by bristy** [comment]
I can confirm that it is still happening. I get following error event though refresh token is valid till `Sunday, May 3, 2026 11:17:13.801 AM` (In 6 months at the time of this message). Does anyone know any workaround?

```
{'error': 'bad_refresh_token', 'error_description': 'The refresh token passed is incorrect or expired.', 'error_uri': 'https://docs.github.com/apps/managing-oauth-apps/troubleshooting-oauth-app-access-token-request-errors/#bad-verification-code'}
```
